File: apps/api/services/identity_sync.py
"""Keeps the identity provider (WorkOS) in step with local lifecycle writes.

Every local user/organization/membership row mirrors a WorkOS record. When a
local record is deleted, the WorkOS record must go with it — otherwise the
provider accumulates accounts that can still authenticate but have no local
counterpart, which is exactly the drift Console surfaces as "out of sync".

The helpers here are deliberately **idempotent**: a record that is already gone
at the provider is treated as success, so a retried delete (or a reconciliation
pass) converges instead of failing.

Ordering contract for callers: write locally, flush, then call these. A provider
failure then raises and rolls the local write back, leaving both systems
untouched. The reverse order can delete at WorkOS and keep the local row.
"""

from __future__ import annotations

from core.errors import AppHTTPException
from core.logging import get_logger
from core.org_permissions import OWNER_ROLE_NAME
from providers.protocol import AuthProvider

logger = get_logger(__name__)

# WorkOS reports "already gone" as 404. Deleting a record that no longer exists
# is the outcome the caller wanted, so it is never an error.
_ALREADY_GONE_STATUS = 404

# WorkOS ships `admin` and `member` roles in every environment. Only the owner
# role is mapped explicitly; everything else takes the environment default so a
# custom 876 org role can never fail the call with an unknown slug.
_PROVIDER_ADMIN_ROLE_SLUG = "admin"


async def ensure_provider_membership(
    provider: AuthProvider,
    *,
    workos_organization_id: str | None,
    workos_user_id: str | None,
    role: str,
) -> str | None:
    """Create the WorkOS organization membership mirroring a local one.

    Returns the WorkOS membership id, or None when either side has no provider
    record to link (a local-only org, e.g. one seeded before WorkOS existed).
    Idempotent: an existing provider membership is adopted rather than
    duplicated, so a retry converges instead of raising a conflict.
    """
    if not workos_organization_id or not workos_user_id:
        logger.info(
            "identity_sync.membership.no_provider_record",
            workos_organization_id=workos_organization_id,
            workos_user_id=workos_user_id,
        )
        return None

    role_slug = _PROVIDER_ADMIN_ROLE_SLUG if role == OWNER_ROLE_NAME else None
    try:
        created = await provider.create_organization_membership(
            user_id=workos_user_id,
            organization_id=workos_organization_id,
            role_slug=role_slug,
        )
    except AppHTTPException:
        existing = await _find_provider_membership(
            provider,
            workos_organization_id=workos_organization_id,
            workos_user_id=workos_user_id,
        )
        if existing is None:
            raise
        logger.info(
            "identity_sync.membership.adopted",
            workos_membership_id=existing,
            workos_organization_id=workos_organization_id,
            workos_user_id=workos_user_id,
        )
        return existing

    membership_id = str(created["id"])
    logger.info(
        "identity_sync.membership.created",
        workos_membership_id=membership_id,
        workos_organization_id=workos_organization_id,
        workos_user_id=workos_user_id,
    )
    return membership_id


async def _find_provider_membership(
    provider: AuthProvider,
    *,
    workos_organization_id: str,
    workos_user_id: str,
) -> str | None:
    memberships = await provider.list_organization_memberships(
        organization_id=workos_organization_id,
        user_id=workos_user_id,
    )
    if not memberships:
        return None
    return str(memberships[0]["id"])


async def delete_provider_user(provider: AuthProvider, workos_user_id: str | None, *, local_user_id: str) -> bool:
    """Delete the WorkOS user backing a local account. Returns True if a call landed."""
    if not workos_user_id:
        logger.info("identity_sync.user.no_provider_record", user_id=local_user_id)
        return False

    try:
        await provider.delete_user(user_id=workos_user_id)
    except AppHTTPException as exc:
        if exc.status_code != _ALREADY_GONE_STATUS:
            raise
        logger.info(
            "identity_sync.user.already_absent",
            user_id=local_user_id,
            workos_user_id=workos_user_id,
        )
        return False

    logger.info("identity_sync.user.deleted", user_id=local_user_id, workos_user_id=workos_user_id)
    return True


async def delete_provider_organization(
    provider: AuthProvider,
    workos_organization_id: str | None,
    *,
    local_organization_id: str,
) -> bool:
    """Delete the WorkOS organization backing a local org. Returns True if a call landed."""
    if not workos_organization_id:
        logger.info("identity_sync.organization.no_provider_record", organization_id=local_organization_id)
        return False

    try:
        await provider.delete_organization(organization_id=workos_organization_id)
    except AppHTTPException as exc:
        if exc.status_code != _ALREADY_GONE_STATUS:
            raise
        logger.info(
            "identity_sync.organization.already_absent",
            organization_id=local_organization_id,
            workos_organization_id=workos_organization_id,
        )
        return False

    logger.info(
        "identity_sync.organization.deleted",
        organization_id=local_organization_id,
        workos_organization_id=workos_organization_id,
    )
    return True


async def delete_provider_membership(
    provider: AuthProvider,
    workos_membership_id: str | None,
    *,
    local_membership_id: str,
) -> bool:
    """Delete the WorkOS organization membership backing a local one."""
    if not workos_membership_id:
        logger.info("identity_sync.membership.no_provider_record", membership_id=local_membership_id)
        return False

    try:
        await provider.delete_organization_membership(membership_id=workos_membership_id)
    except AppHTTPException as exc:
        if exc.status_code != _ALREADY_GONE_STATUS:
            raise
        logger.info(
            "identity_sync.membership.already_absent",
            membership_id=local_membership_id,
            workos_membership_id=workos_membership_id,
        )
        return False

    logger.info(
        "identity_sync.membership.deleted",
        membership_id=local_membership_id,
        workos_membership_id=workos_membership_id,
    )
    return True


async def compensate_provider_user(provider: AuthProvider, workos_user_id: str, *, operation: str) -> None:
    """Best-effort rollback of a WorkOS user created earlier in a failed request.

    Never raises: it runs inside an `except` block that is about to re-raise the
    original failure, and masking that with a compensation error would hide the
    real cause. A compensation that fails leaves an orphan for the reconciler.
    """
    try:
        await provider.delete_user(user_id=workos_user_id)
        logger.info("identity_sync.compensated", resource="user", operation=operation, workos_user_id=workos_user_id)
    except Exception:
        logger.warning(
            "identity_sync.compensation_failed",
            resource="user",
            operation=operation,
            workos_user_id=workos_user_id,
            exc_info=True,
        )

File: apps/api/src/services/identity-sync.ts
import { isAppHttpError } from '@/http/errors'
import { getLogger } from '@/platform/logger'
import { OWNER_ROLE_NAME } from '@/platform/permissions'

/**
 * Keeps the identity provider (WorkOS) in step with local lifecycle writes.
 *
 * Every local user, organization, and membership row mirrors a WorkOS record.
 * When a local record goes, the provider record must go with it — otherwise
 * WorkOS accumulates accounts that can still authenticate but have no local
 * counterpart, which is the drift Console surfaces as "out of sync".
 *
 * The helpers are retry-safe where the provider exposes an idempotent outcome:
 * a record already gone at WorkOS is treated as success. Callers choose ordering
 * based on the safety property of the mutation. Membership role changes update
 * the provider first so a demotion cannot leave an elevated WorkOS role behind;
 * deletes remove local access first, then deprovision the provider record.
 */

const log = getLogger('identity-sync')

/** WorkOS reports "already gone" as 404, which is the outcome the caller wanted. */
const ALREADY_GONE_STATUS = 404

/**
 * WorkOS ships `admin` and `member` in every environment. The 876 owner role
 * maps to provider admin; all other/custom org roles map to provider member so
 * application-specific authorization stays in the 876 data plane.
 */
const PROVIDER_ADMIN_ROLE_SLUG = 'admin'
const PROVIDER_MEMBER_ROLE_SLUG = 'member'

function providerRoleSlug(role: string): string {
  return role === OWNER_ROLE_NAME
    ? PROVIDER_ADMIN_ROLE_SLUG
    : PROVIDER_MEMBER_ROLE_SLUG
}

/** The provider surface these helpers need — not the whole adapter. */
export type IdentitySyncProvider = {
  createOrganizationMembership(params: {
    userId: string
    organizationId: string
    roleSlug?: string | null
  }): Promise<Record<string, unknown>>
  updateOrganizationMembership(
    membershipId: string,
    params: { roleSlug: string }
  ): Promise<Record<string, unknown>>
  listOrganizationMemberships(filters: {
    organizationId?: string | null
    userId?: string | null
  }): Promise<Record<string, unknown>[]>
  deleteOrganizationMembership(membershipId: string): Promise<void>
  deleteUser(userId: string): Promise<void>
  deleteOrganization(organizationId: string): Promise<void>
}

function isAlreadyGone(error: unknown): boolean {
  return isAppHttpError(error) && error.httpStatus === ALREADY_GONE_STATUS
}

/**
 * Create the WorkOS membership mirroring a local one.
 *
 * Returns the provider membership id, or `null` when either side has no
 * provider record to link — a local-only org, for instance one seeded before
 * WorkOS existed. An existing provider membership is **adopted** rather than
 * duplicated, so a retry converges instead of raising a conflict.
 */
export async function ensureProviderMembership(
  provider: IdentitySyncProvider,
  params: {
    workosOrganizationId: string | null
    workosUserId: string | null
    role: string
  }
): Promise<string | null> {
  if (!params.workosOrganizationId || !params.workosUserId) {
    log.info(
      {
        workos_organization_id: params.workosOrganizationId,
        workos_user_id: params.workosUserId,
      },
      'identity_sync.membership.no_provider_record'
    )
    return null
  }

  // Creation intentionally omits the role for non-owners so WorkOS can apply
  // the environment's configured default. Owner must be elevated explicitly.
  const roleSlug =
    params.role === OWNER_ROLE_NAME ? PROVIDER_ADMIN_ROLE_SLUG : null

  let created: Record<string, unknown>
  try {
    created = await provider.createOrganizationMembership({
      userId: params.workosUserId,
      organizationId: params.workosOrganizationId,
      roleSlug,
    })
  } catch (error) {
    const existing = await findProviderMembership(provider, {
      workosOrganizationId: params.workosOrganizationId,
      workosUserId: params.workosUserId,
    })
    if (existing === null) throw error

    log.info(
      {
        workos_membership_id: existing,
        workos_organization_id: params.workosOrganizationId,
        workos_user_id: params.workosUserId,
      },
      'identity_sync.membership.adopted'
    )
    return existing
  }

  const membershipId = String(created.id)
  log.info(
    {
      workos_membership_id: membershipId,
      workos_organization_id: params.workosOrganizationId,
      workos_user_id: params.workosUserId,
    },
    'identity_sync.membership.created'
  )
  return membershipId
}

async function findProviderMembership(
  provider: IdentitySyncProvider,
  params: { workosOrganizationId: string; workosUserId: string }
): Promise<string | null> {
  const memberships = await provider.listOrganizationMemberships({
    organizationId: params.workosOrganizationId,
    userId: params.workosUserId,
  })

  return memberships.length > 0 ? String(memberships[0]?.id) : null
}

/**
 * Synchronize the provider role for an existing local membership.
 *
 * A missing provider id/record is safe to tolerate: there is no provider-side
 * privilege left to preserve. Other provider failures are re-thrown so callers
 * can avoid committing a local role change that WorkOS rejected.
 */
export async function updateProviderMembershipRole(
  provider: IdentitySyncProvider,
  workosMembershipId: string | null,
  role: string,
  params: { localMembershipId: string }
): Promise<boolean> {
  if (!workosMembershipId) {
    log.info(
      { membership_id: params.localMembershipId },
      'identity_sync.membership.no_provider_record'
    )
    return false
  }

  const roleSlug = providerRoleSlug(role)
  try {
    await provider.updateOrganizationMembership(workosMembershipId, {
      roleSlug,
    })
  } catch (error) {
    if (!isAlreadyGone(error)) throw error

    log.info(
      {
        membership_id: params.localMembershipId,
        workos_membership_id: workosMembershipId,
      },
      'identity_sync.membership.already_absent'
    )
    return false
  }

  log.info(
    {
      membership_id: params.localMembershipId,
      workos_membership_id: workosMembershipId,
      role_slug: roleSlug,
    },
    'identity_sync.membership.role_updated'
  )
  return true
}

/** Delete the WorkOS user backing a local account. True when a call landed. */
export async function deleteProviderUser(
  provider: IdentitySyncProvider,
  workosUserId: string | null,
  params: { localUserId: string }
): Promise<boolean> {
  if (!workosUserId) {
    log.info(
      { user_id: params.localUserId },
      'identity_sync.user.no_provider_record'
    )
    return false
  }

  try {
    await provider.deleteUser(workosUserId)
  } catch (error) {
    if (!isAlreadyGone(error)) throw error

    log.info(
      { user_id: params.localUserId, workos_user_id: workosUserId },
      'identity_sync.user.already_absent'
    )
    return false
  }

  log.info(
    { user_id: params.localUserId, workos_user_id: workosUserId },
    'identity_sync.user.deleted'
  )
  return true
}

/** Delete the WorkOS organization backing a local org. True when a call landed. */
export async function deleteProviderOrganization(
  provider: IdentitySyncProvider,
  workosOrganizationId: string | null,
  params: { localOrganizationId: string }
): Promise<boolean> {
  if (!workosOrganizationId) {
    log.info(
      { organization_id: params.localOrganizationId },
      'identity_sync.organization.no_provider_record'
    )
    return false
  }

  try {
    await provider.deleteOrganization(workosOrganizationId)
  } catch (error) {
    if (!isAlreadyGone(error)) throw error

    log.info(
      {
        organization_id: params.localOrganizationId,
        workos_organization_id: workosOrganizationId,
      },
      'identity_sync.organization.already_absent'
    )
    return false
  }

  log.info(
    {
      organization_id: params.localOrganizationId,
      workos_organization_id: workosOrganizationId,
    },
    'identity_sync.organization.deleted'
  )
  return true
}

/** Delete the WorkOS membership backing a local one. True when a call landed. */
export async function deleteProviderMembership(
  provider: IdentitySyncProvider,
  workosMembershipId: string | null,
  params: { localMembershipId: string }
): Promise<boolean> {
  if (!workosMembershipId) {
    log.info(
      { membership_id: params.localMembershipId },
      'identity_sync.membership.no_provider_record'
    )
    return false
  }

  try {
    await provider.deleteOrganizationMembership(workosMembershipId)
  } catch (error) {
    if (!isAlreadyGone(error)) throw error

    log.info(
      {
        membership_id: params.localMembershipId,
        workos_membership_id: workosMembershipId,
      },
      'identity_sync.membership.already_absent'
    )
    return false
  }

  log.info(
    {
      membership_id: params.localMembershipId,
      workos_membership_id: workosMembershipId,
    },
    'identity_sync.membership.deleted'
  )
  return true
}

/**
 * Best-effort rollback of a WorkOS user created earlier in a failed request.
 *
 * **Never throws.** It runs inside a `catch` that is about to re-raise the
 * original failure, and letting a compensation error escape would replace the
 * real cause with a confusing one. A compensation that fails leaves an orphan
 * for the reconciler, which is the lesser problem.
 */
export async function compensateProviderUser(
  provider: IdentitySyncProvider,
  workosUserId: string,
  params: { operation: string }
): Promise<void> {
  try {
    await provider.deleteUser(workosUserId)
    log.info(
      {
        resource: 'user',
        operation: params.operation,
        workos_user_id: workosUserId,
      },
      'identity_sync.compensated'
    )
  } catch (error) {
    log.warn(
      {
        err: error,
        resource: 'user',
        operation: params.operation,
        workos_user_id: workosUserId,
      },
      'identity_sync.compensation_failed'
    )
  }
}
